### PR TITLE
JW-295 Added a lot of ASSERTs and removed some redundant code

### PR DESCRIPTION
**Pull request by Attil**

### CODE REVIEW

**Comment by m-kostrzewa** [comment]
I'm assigning someone else to review

File: windows/vr_host.c [hunk 1]
```diff
@@ -31,20 +33,23 @@ NDIS_IO_WORKITEM_FUNCTION deferred_work_routine;
 static void win_pfree(struct vr_packet *pkt, unsigned short reason);  // Forward declaration
 
 static NDIS_STATUS
-create_forwarding_context(PNET_BUFFER_LIST nbl)
+create_forwarding_context(NET_BUFFER_LIST* nbl)
```
Comment (michal-clapinski) [inline windows/vr_host.c]: why did you change this?

File: windows/vrouter_mod.c [hunk 2]
```diff
@@ -22,7 +23,7 @@ ULONG  SxExtOidRequestId = 'RVCO';
 PSX_SWITCH_OBJECT SxSwitchObject = NULL;
 NDIS_HANDLE SxNBLPool = NULL;
 
-unsigned int vr_num_cpus = 1;
+unsigned int vr_num_cpus;
```
Comment (michal-clapinski) [inline windows/vrouter_mod.c]: Why did you change this? It is done this way on freebsd and linux if I recall correctly.

File: windows/vr_host.c (hunk 1 above)
Comment (Attil) [inline windows/vr_host.c]: I replaced all `PNET_BUFFER_LIST` in vr_host.c with `NET_BUFFER_LIST*`, because it is more readable this way I think. Additionally, the rest of the project uses `*` for marking a pointer instead of typedefing the type once more as `Ptype`.

Comment (m-kostrzewa) [inline windows/vr_host.c]: I agree this is more readable this way, but doesn't this clash with "standard" windows kernel conventions?

File: windows/vrouter_mod.c (hunk 2 above)
Comment (Attil) [inline windows/vrouter_mod.c]: There is no point to this IMHO. We do it first think ever, at the start of initialization routine.

File: windows/vr_host.c (hunk 1 above)
Comment (Attil) [inline windows/vr_host.c]: Standard windows kernel conventions are in contradiction to existing vRouter conventions and possibly C conventions. I've decided consistency is the most important factor here. Don't you agree?

Comment (michal-clapinski) [inline windows/vr_host.c]: I don't agree. I like when platform independent code looks like standard C code and Windows code looks like code for Windows. Though I agree it is more readable with ```*```.

**Comment by Attil** [comment]
Bump

File: windows/vr_host.c [hunk 3]
```diff
@@ -221,6 +249,9 @@ win_page_free(void *address, unsigned int size)
 struct vr_packet *
 win_get_packet(PNET_BUFFER_LIST nbl, struct vr_interface *vif, unsigned char flags)
 {
+    ASSERT(nbl != NULL);
+    ASSERTMSG("No source provided", (flags & (VP_WIN_CLONED | VP_WIN_CREATED | VP_WIN_RECEIVED)) == 0);
```
Comment (michal-clapinski) [inline windows/vr_host.c]: On line 126 in this file you used a variable for these flags. Choose one solution and be consistent.

File: windows/vrouter_mod.c [hunk 4]
```diff
@@ -6,6 +6,7 @@
 #include "vr_sandesh.h"
 
 extern void vif_attach(struct vr_interface *vif);
+extern void vif_detach(struct vr_interface *vif);
```
Comment (michal-clapinski) [inline windows/vrouter_mod.c]: I don't like this. Can't you include some header?

File: windows/vr_host.c [hunk 5]
```diff
@@ -7,6 +7,8 @@
 #include "vrouter.h"
 
 extern void vif_attach(struct vr_interface *vif);
+extern void vr_malloc_stats(unsigned int, unsigned int);
+extern void vr_free_stats(unsigned int);
```
Comment (michal-clapinski) [inline windows/vr_host.c]: Please use headers.

File: windows/vrouter_mod.c (hunk 4 above)
Comment (Attil) [inline windows/vrouter_mod.c]: I can do that, but no header file declares this symbol :( It's used as extern in Linux version, but I agree it's a bad practice.

**Comment by michal-clapinski** [comment]
Retest this please

**Comment by Attil** [comment]
Retest this please

**Comment by michal-clapinski** [comment]
I'll review this after rebase

File: include/vr_stats.h [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+#endif
+
+extern void vr_malloc_stats(unsigned int, unsigned int);
+extern void vr_free_stats(unsigned int);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
```
Comment (michal-clapinski) [inline include/vr_stats.h]: new line char missing

File: windows/vr_host.c [hunk 7]
```diff
@@ -100,32 +106,41 @@ delete_created_nbl(PNET_BUFFER_LIST nbl)
...
-        NDIS_SEND_COMPLETE_FLAGS_SWITCH_SINGLE_SOURCE);
+        NDIS_SEND_COMPLETE_FLAGS_SWITCH_SINGLE_SOURCE | NDIS_SEND_FLAGS_SWITCH_DESTINATION_GROUP);
 }
 
 static void
 free_associated_nbl(struct vr_packet* pkt)
 {
+    ASSERT(pkt != NULL);
+
+    const unsigned char any_flag = VP_WIN_RECEIVED | VP_WIN_CLONED | VP_WIN_CREATED;
```
Comment (michal-clapinski) [inline windows/vr_host.c]: make it global

File: windows/vr_host.c [hunk 8]
```diff
@@ -376,6 +417,11 @@ win_pexpand_head(struct vr_packet *pkt, unsigned int hspace)
 static void
 win_preset(struct vr_packet *pkt)
 {
+    ASSERT(pkt != NULL);
+
+    if (!pkt)
+        return;
```
Comment (michal-clapinski) [inline windows/vr_host.c]: Seems useless after the assert

File: windows/vr_host.c [hunk 9]
```diff
@@ -1082,10 +1139,13 @@ vrouter_generate_pool()
...
+    ASSERT(pool != NULL);
+
     return pool;
 }
 
 void vrouter_free_pool(NDIS_HANDLE pool)
 {
+    ASSERTMSG("NBL pool is not initialized", pool != NULL);
     NdisFreeNetBufferListPool(pool);
 }
```
Comment (michal-clapinski) [inline windows/vr_host.c]: new line char missing

**Comment by Attil** [comment]
Bump.
